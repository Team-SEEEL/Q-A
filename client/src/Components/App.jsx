/* eslint-disable import/extensions */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import axios from 'axios';
import styled from 'styled-components';
import SearchForm from './searchForm.jsx';
import AnsweredQuestions from './answeredQuestions.jsx';
import NavTabs from './navigationTabs.jsx';
import PostQuestion from './postQuestion.jsx';

const StyledContainer = styled.div`
  border-top : 1px solid grey;
  display: flex;
  flex-direction: column;
  width: 100%;
`;

const HighLight = styled.span`
  background-color: yellow;
`;

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      text: '',
      view: 'home',
      questions: [],
      filteredQuestions: [],
      page: '',
    };
    this.searchQuestions = this.searchQuestions.bind(this);
  }

  componentDidMount() {
    let index;
    if (window.location.pathname === '/') {
      index = Math.floor(Math.random() * 100);
    } else {
      index = parseInt(window.location.pathname.slice(1));
    }
    axios.get(`/questions/api/products/${index}`)
      .then((response) => {
        this.setState({ questions: response.data, page: index });
      })
      .catch((err) => {
        console.log(err);
      });
  }

  searchQuestions(query) {
    this.setState({ filteredQuestions: [] });
    let searchResults = [...this.state.questions];
    query = query.toLowerCase();
    const { questions } = this.state;
    searchResults = searchResults.filter((question) => question.answers.every((answer) => answer.answer.toLowerCase().includes(query)) || question.body.toLowerCase().includes(query));
    this.setState({ filteredQuestions: searchResults, view: 'search', text: query });
  }

  render() {
    if (this.state.view === 'search') {
      return (
        <div className="main-container">
          <StyledContainer>
            <h2>Customer questions & answers</h2>
            <SearchForm search={this.searchQuestions} value={this.state.text} />
            <NavTabs />
            <AnsweredQuestions questions={this.state.filteredQuestions} searched={this.state.text} />
            <PostQuestion post={this.postQuestion} />
          </StyledContainer>
        </div>
      );
    }
    return (
      <div className="main-container">
        <StyledContainer>
          <h2>Customer questions & answers</h2>
          <SearchForm search={this.searchQuestions} value={this.state.text} />
          <AnsweredQuestions questions={this.state.questions} />
          <PostQuestion post={this.postQuestion} page={this.state.page} />
        </StyledContainer>
      </div>
    );
  }
}

export default App;
