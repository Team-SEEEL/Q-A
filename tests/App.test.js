/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-undef */
import App from '../client/src/Components/App';

const axios = require('axios');

describe('<App />', () => {
  // it('Should return questions with a GET request', () => {
  //   return axios.get('http://localhost:3000/api/questions')
  //     .then((response) => {
  //       expect(response.data).toHaveLength(100);
  //     });
  // });

  it('Should render correctly', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  xit('Should render 12 question components', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(App)).to.have.lengthOf(12);
  });

  xit('Should render children when passed in', () => {
    const wrapper = shallow((
      <App>
        <div className="unique" />
      </App>
    ));
    expect(wrapper.contains(<div className="unique" />).to.equal(true));
  });

  xit('Should simulate click events', () => {
    const onButtonClick = sinon.spy();
    const wrapper = shallow(<App onButtonClick={onButtonClick} />);
    wrapper.find('button').simulate('click');
    expect(onButtonClick).to.have.property('callCount', 1);
  });

  xit('Should simulate typing', () => {
    const onKeystroke = sinon.spy();
    const wrapper = shallow(<App onKeystroke={onKeystroke} />);
    wrapper.find('form').simulate('change');
    expect(onKeystroke).to.have.property('callCount', 1);
  });

  xit('Should change state when onKeystroke() is called', () => {
    const wrapper = shallow(<App />);
    wrapper.instance().onKeystroke('event');
    expect(wrapper.state()).toEqual({ text: 'event' });
  });

  it('Should update a state in a callback', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ question: 'new' }, () => {
      expect(wrapper.state()).toEqual({ question: 'new', text: '' });
    });
  });
});
