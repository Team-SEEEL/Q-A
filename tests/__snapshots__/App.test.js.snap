// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Should render correctly 1`] = `
<div
  className="main-container"
>
  <form>
    Have a question? Search for answers
  </form>
</div>
`;
